src/NUnitEngine/nunit.engine/Runners/AggregatingTestRunner.cs

```diff
@@ -210,16 +227,40 @@ protected override void Dispose(bool disposing)
             base.Dispose(disposing);
 
             foreach (var runner in Runners)
-                runner.Dispose();
+            {
+                try
+                {
+                    runner.Dispose();
+                }
+                catch (NUnitEngineException e)
+                {
+                    _thrownExceptions.Add(e);
+                }
+            }
 
             Runners.Clear();
+            throw new NUnitEngineException(_thrownExceptions);
```

Comment:
Just come across this - D'oh! Of course, that should only throw when there are exceptions to be thrown - will fix this before merge.

Reply:
So, throwing is the whole point of this change. I want to go back and question why we are  throwing in the case of failure to unload. I mentioned this elsewhere (on the issue I think) and I wonder why we can't just report this as an error (or warning) associated with the individual assembly as we do in other cases. The difference here, of course, is that only the engine can detect the error as opposed to the framework. Worth discussing?

Reply:
For the avoidance of doubt, to be clear of the things I'm aiming to do here:

1. Bring the AggregatingTestRunner inline with the other test runners.
2. Have the engine throw an exception on domain unload failure, caught by the runner.

The current issue is that the AggregatingTestRunner is throwing the exception mid-run, where as the other TestRunners would all throw the exception at the end of the test run, as the runner is disposed, and once the full results have already been returned to the runner.

I think that design currently works. A few reasons:
1. I don't think we should silently swallow the exception, this exception indicates an unclean shut down in the user's code. By returning a distinct unload-failure-error-code such as in #320 - we still give the user the _choice_ to ignore this if they wish to.
2. Re: associating with xml - by the time the domain is unloaded and the engine cleaned up, the results have already been returned to the runner. Would we want to alter this behaviour in all situations - e.g. with a GUI, unload wouldn't happen after each run, correct? I don't see the advantage of holding on to the xml, incase we wish to add an error that's not strictly related to the test results.
3. I feel like app domain unload belongs semantically as an engine exception, rather than a 'test result' in the xml. i.e. I could chose to run in the main domain, in which case I would get no exception.

I _don't_ think that this exception should escape the runner - the current behaviour of handling this within the runner seems correct to me.

Reply:
(The term 'runner' is overloaded in this conversation! I've edited to be clear when I mean ITestEngineRunner - and when I mean runner as in 'console runner'/GUI etc.)

Reply:
I see your point about the separation of running from unloading. In that case, I suppose it's up to the runner to decide what to do with the exceptions - error, warn, etc.  That would mean that unload exceptions need to be distinguishable from others of course.

Reply:
How about an NUnitUnloadException which inherits from NUnitEngineExceptions, to not break any existing runners by throwning a new Exception type? 🙂